### PR TITLE
Fix test order (1) - Never write into Fixtures dir

### PR DESCRIPTION
**Pull request by m-vo**
| Q                | A
| -----------------| ---
| Fixed issues     | partly solves #2316 
| Docs PR or issue | -

I'm working on test isolation. This part makes sure tests aren't reading temporary data from each other. This effectively means that a test should never write into the `Fixtures` dir (or anywhere else inside the source directories). If temporary files need to be created this now will happen inside the test's own `tmp` dir.

While working on these files, I'm also adjusting the following for consistency:

1) Refactoring path operations to use `Path`
1) Removing private property `$filesystem`  (if we want to share the `Filesystem` instance across tests, we should imo add it to `ContaoTestCase` as a static property)
1) Use `$filesystem` instead of `$fs` everywhere

Following changes needed for #2316 (but unrelated to writing files) are also part of this PR:

1) Reset template loader every time (was missing once, now in `tearDown`).
1) Do not share `ContaoCacheWarmer` instance.

### CODE REVIEW

**Comment by m-vo** [comment]
I had to backport the changes to `StringUtil#stripRootDir` from #1918 which fixed lots of tests on Windows. 

The ones that are still failing right now are a bit of a mystery for me, though. Maybe I'll have to check this PR out on windows and debug there. Or maybe someone else directly sees the problem? /cc @fritzmg :slightly_smiling_face:

**Comment by leofeyer** [comment]
> 1. Refactoring path operations to use `Path`

Ok if the changes go into Contao 4.11 or 4.12 but not in Contao 4.9.

> 2. Removing private property `$filesystem`  (if we want to share the `Filesystem` instance across tests, we should imo add it to `ContaoTestCase` as a static property)

✅ 

> 3. Use `$filesystem` instead of `$fs` everywhere

I fail to see the improvement here…

> 1. Reset template loader every time (was missing once, now in `tearDown`).
> 2. Do not share `ContaoCacheWarmer` instance.

✅ 

> I had to backport the changes to StringUtil#stripRootDir from #1918

Another reason why we should probably only change this in Contao 4.11.

File: core-bundle/src/Resources/contao/library/Contao/StringUtil.php [hunk 1]
```diff
@@ -1112,10 +1113,12 @@ public static function trimsplit($strPattern, $strString)
...
 	public static function stripRootDir($path)
 	{
-		$projectDir = System::getContainer()->getParameter('kernel.project_dir');
+		// Compare normalized version of the paths
+		$projectDir = Path::normalize(System::getContainer()->getParameter('kernel.project_dir'));
+		$normalizedPath = Path::normalize($path);
 		$length = \strlen($projectDir);
 
-		if (strncmp($path, $projectDir, $length) !== 0 || \strlen($path) <= $length || ($path[$length] !== '/' && $path[$length] !== '\\'))
+		if (strncmp($normalizedPath, $projectDir, $length) !== 0 || \strlen($normalizedPath) <= $length || $normalizedPath[$length] !== '/')
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/StringUtil.php:1121]: This change seems unrelated. Am I missing something?

Also – while we are at it – why are you not using `Path::isBasePath()`?

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/library/Contao/StringUtil.php (hunk 1 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/StringUtil.php:1121]: It's needed because due to the normalization `\` &rarr; `/` together with legacy code paths using `\`, stripping the root dir could break on windows (analogous to #1902, https://github.com/contao/contao/pull/1902/commits/48b1f8998aca6a23c268927d6bf4666893421cb6).

 See this CI run: https://github.com/contao/contao/pull/2516/checks?check_run_id=2088691955

> Also – while we are at it – why are you not using Path::isBasePath()?

There are some edge cases to this. We fixed the same issue in https://github.com/contao/contao/pull/1918. Now the implementation should be identical to the upstream versions.

**Comment by m-vo** [comment]
Thanks for fixing tests on windows @fritzmg! :heart:

**Comment by fritzmg** [comment]
Never underestimate the value of Windows! There at least some errors don't fail silently as they do in *nix 😁 (regarding symlinks).

**Review by fritzmg — APPROVED** [review]
I am very much in favour of merging this PR into `4.9`, including all `Path` related refactors. It has been shown time and time again that _not_ using it leads to issues.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.
